Playwright: add unit tests to `RestAPIClient`.

worldomonation opened the pull request:
#### Proposed Changes

This PR adds unit tests to the `RestAPIClient` module within `@automattic/calypso-e2e` as part of a 20% time project.

Key changes:
- additional data stored in the SecretsManager > testAccounts;
- new file containing unit tests for private and public methods of RestAPIClient;
- improvements to behavior of some methods in RestAPIClient.

#### Testing Instructions

Ensure the following:
  - [x] Gutenberg E2E (desktop)
  - [x] Gutenberg E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E
  - [x] Check Code Style
  - [x] Unit Tests

Related to #

worldomonation committed:
'Add unit tests for `getMyAccountInformation`

worldomonation commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -404,6 +404,9 @@ object RunAllUnitTests : BuildType({
 				unset CALYPSO_ENV
 
+				# Decrypt calypso-e2e secrets
+				E2E_SECRETS_KEY="%E2E_SECRETS_ENCRYPTION_KEY_CURRENT%" yarn workspace @automattic/calypso-e2e decrypt-secrets
```
This PR requires interaction with the `calypso-e2e` secrets manager.

dpasque replied:
So, I feel pretty strongly that we should not add this dependency to this build. Or, put otherwise, we should just mock the `.secrets` get property on the `SecretsManager` in the unit tests. Otherwise, we're adding a pretty significant dependency to what are otherwise pure tests.

This also means people will have to encrypt the E2E secrets file on their machine if they want to run unit tests, and that feels like it could be pretty disruptive.

What do you think?

worldomonation replied:
> This also means people will have to encrypt the E2E secrets file on their machine if they want to run unit tests, and that feels like it could be pretty disruptive.

I hadn't considered that possibility but I think you are correct that it would require developers to _decrypt_ the secrets, but only if they're running the entire suit of tests under `packages/` - which is an unlikely proposition, but one that's not entirely impossible I admit.

Reason I say that this is unlikely is because I imagine that developers would typically like to use the `watch` variant so that each change don't trigger a 10-minute test run.

In either case, as you suggested in a subsequent review to eliminate the integration part of these tests, I think it would make sense to mock the `SecretsManager` instead.

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -74,13 +77,20 @@ export class RestAPIClient {
+			new URL( BEARER_TOKEN_URL ),
+			{
+				method: 'post',
+				body: params,
+			}
+		);
+
+		if ( response.success === false ) {
+			const firstError = response.data.errors.at( 0 );
+			throw new Error( `${ firstError?.code }: ${ firstError?.message }` );
+		}
 
-		this.bearerToken = response.data.bearer_token;
-		return this.bearerToken;
+		return response.data.bearer_token;
```
(Completely optional/stylistic) -- can you explain more about moving this out of this method? Since this effectively a caching operation, I think it makes a lot of sense to have it in the method that fetches the token (as opposed to the function that constructs the request header)

worldomonation replied:
So, I changed the signature to only return the bearer token because the `getBearerToken` method's name suggests its purpose, which is to return the bearer token. What the caller function does with the token is not of interest to the `getBearerToken` method.

This change was also made to support this attempt at unit testing the class. Having the side effect of setting the bearer token required network call to the authentication endpoint, which was not desirable from the unit test perspective when the intent was not to test the actual process of retrieving the bearer token.

dpasque commented on packages/calypso-e2e/src/test/rest-api-client.test.ts:
```diff
@@ -0,0 +1,137 @@
+			username: credentials.username,
+			password: credentials.password,
+		} );
+		const bearerToken = await ( restAPIClient as any ).getBearerToken();
```
In general, I tend to agree with the philosophy that we should only unit test the public methods/interface of a piece of code. 

I think there are three risks that come with unit testing private methods...
1. You aren't testing your code in the same way it's being called, which makes it less honest to the consumer experience.
2. Your tests are more fragile to refactors over time. If tests only target the public methods, you are free to refactor the private methods and implementations as much as you want and can still re-use the same tests!
3. You often have to do less-ideal things like dropping type safety/overriding es-lint rules, etc.

I realize that's a pretty disruptive change for this PR, so it's not required, but I would highly recommend it!

Also, I could see `getBearerToken` also getting exposed as a public method too -- it seems like a request that people may want to make in isolation for other purposes!

worldomonation replied:
Whew, this is a lot to unpack in a comment! 😄 

Generally, I think you raise good points and the assertions are backed up on numerous sources [1](https://softwareengineering.stackexchange.com/questions/100959/how-do-you-unit-test-private-methods)[2](https://anthonysciamanna.com/2016/02/14/should-private-methods-be-tested.html). 

> You often have to do less-ideal things like dropping type safety/overriding es-lint rules, etc.

This is also true and you can see that play out in this test as well.

On the other hand, I did find numerous issues with how error/unexpected cases were handled for private methods such as `getBearerToken` and `getRequestURL`. For instance, only through writing the unit tests for `getBearerToken` did I find out that if the response JSON was not in the expected format of `BearerTokenResponse` the class property and returned value would both be `undefined`. 

I would imagine unit tests written following the "do not test private methods" would have masked the `getBearerToken` issue described above, resulting in hours spent identifying the actual root cause. 

Now, the last suggestion - to make `getBearerToken` public - let's do it. I originally wanted to hide this method as I felt callers had no business being able to call this manually, but it would make sense that it can be useful for other purposes.

dpasque replied:
>Whew, this is a lot to unpack in a comment! 😄

Ack, sorry! Didn't mean to throw too much all at once! 😅 

>On the other hand, I did find numerous issues with how error/unexpected cases were handled for private methods such as getBearerToken and getRequestURL. For instance, only through writing the unit tests for getBearerToken did I find out that if the response JSON was not in the expected format of BearerTokenResponse the class property and returned value would both be undefined.
I would imagine unit tests written following the "do not test private methods" would have masked the getBearerToken issue described above, resulting in hours spent identifying the actual root cause.

That's an interesting point! I would actually be surprised though if that failure was masked in testing at the other public request seams. One important clarification I should make is that I don't think we should avoid testing the _content and consequences_ of private methods, we should just try to test them using the public methods as the seams.

So, for example, you could test the bearer token handling specifically in its own test, and do so by calling another public method, and using `nock` validation to ensure that a valid bearer token was attached to the request. I think that test would have failed with an undefined value in that header, which would hopefully lead to quick debugging on the `getBearerToken` method.

I may be misunderstanding the failure though! And this is definitely just a matter of opinion 😁  (and tradeoffs too!)

dpasque commented on packages/calypso-e2e/src/test/rest-api-client.test.ts:
```diff
@@ -0,0 +1,137 @@
+		} );
+		const bearerToken = await ( restAPIClient as any ).getBearerToken();
+
+		expect( bearerToken ).toBeDefined();
+		expect( typeof bearerToken ).toBe( 'string' );
+		expect( bearerToken.length ).toBe( 64 );
```
Because these are actually making web requests over the network, I think these qualify as integration tests. Our unit tests I believe have the assumption that they should be able to run entirely in-memory on a machine without network.

With that, I think we probably need to either move these tests out of the unit test configs/build to an integration test config/build, or we should mock the network traffic. There are a couple of good libraries for doing that, [nock](https://www.npmjs.com/package/nock), which is already used in the monorepo, and [msw](https://www.npmjs.com/package/msw), which is not in the monorepo, but seems to have a nicer API.

worldomonation replied:
Fair point, it would have been better to call these integration tests. 

AFAIK there isn't an integration build for `wp-calypso` currently, so I'll go ahead and mock the network calls for now. The same will apply to other tests that make actual network calls.

dpasque reviewed (changes requested):
Yay for testing the Rest API client! 🎉  I think that's an awesome idea!

I have a few bigger-picture questions/requests, let me know what you think or if you have any follow-up questions! 👍

worldomonation committed:
Revise the PR:
- use nock to mock responses;
- remove the use of actual network requests from the unit tests;
- change the interface definition of AllSitesResponse to use generics;
- remove the key decryption step from Unit Tests;

dpasque commented on packages/calypso-e2e/src/test/rest-api-client.test.ts:
```diff
@@ -0,0 +1,179 @@
+		} );
+		const bearerToken = await restAPIClient.getBearerToken();
+
+		expect( bearerToken ).toBeDefined();
+		expect( bearerToken ).toBe( mockedToken );
+		expect( typeof bearerToken ).toBe( 'string' );
```
(optional small nit) -- I think you might only need the `toBe` check here. AFAIK, it's a referential comparison, so if `bearerToken` is not a string or is undefined, that check should fail!

worldomonation replied:
It doesn't look like this is the case:

```
    Expected: "string"
    Received: "abcdefghijklmn"

      47 | 		expect( bearerToken ).toBeDefined();
      48 | 		expect( bearerToken ).toBe( mockedToken );
    > 49 | 		expect( bearerToken ).toBe( 'string' );
```

dpasque reviewed (approved):
These unit tests look beautiful! 😍  Thanks for making all the updates! Left one small optional nit, but otherwise LGTM! 🚀 

(Also, something looks amiss with the E2Es, I wonder if there was a recent merge that broke that package, and that came in through the rebase? I'll check that out too.)

dpasque commented:
⚠️ Ah wait, nope, I think you need to add nock and maybe its types to the `calypso-e2e` package -- it's failing to build!

`src/test/rest-api-client.test.ts(2,18): error TS2307: Cannot find module 'nock' or its corresponding type declarations.`

This should be done before merging!

worldomonation commented:
> ⚠️ Ah wait, nope, I think you need to add nock and maybe its types to the `calypso-e2e` package -- it's failing to build!
> 
> `src/test/rest-api-client.test.ts(2,18): error TS2307: Cannot find module 'nock' or its corresponding type declarations.`
> 
> This should be done before merging!

Yep, I had been doing something else and figured out it was the `nock` package but hadn't gotten around to fixing it yet. Should be able to get that done today.

worldomonation committed:
Add `nock` to calypso-e2e.
